crates/solver/src/driver/solver_settlements.rs

```diff
@@ -90,30 +93,138 @@ fn merge_at_most_settlements(
     }
 }
 
-pub fn filter_settlements_without_old_orders(
-    min_order_age: Duration,
-    settlements: &mut Vec<Settlement>,
-) {
+/// Filters out all settlements without any orders which are mature by age or mature by association.
+/// Any order older than `min_order_age` is considered to be mature by age.
+/// Any younger order in a settlement containing an order mature by age or mature by association
+/// is considered to be mature by association.
+/// Because maturity by association is defined recursively it can "spread" across settlements,
+/// resulting in settlements being allowed where it's not immediately obvious by which association
+/// any order of a settlement has matured.
+pub fn filter_immature_settlements(min_order_age: Duration, settlements: &mut Vec<Settlement>) {
+    use model::order::Order;
     let settle_orders_older_than =
         chrono::offset::Utc::now() - chrono::Duration::from_std(min_order_age).unwrap();
-    settlements.retain(|settlement| {
-        settlement
-            .trades()
-            .iter()
-            .any(|trade| trade.order.order_meta_data.creation_date <= settle_orders_older_than)
-    });
+
+    let mut valid_trades = HashSet::<&Order>::default();
+    // Use *const T to avoid expensive clones and hashing. This also doesn't require T: Hash + Eq.
+    // This approach is correct as long as 'settlements' doesn't reallocate.
+    let mut valid_settlements = HashSet::<*const Settlement>::default();
```

Comment:
I think one more condition on correctness here is that we don't mutate the `settlements: &mut Vec<Settlement>`. For example, something as innocent as `settlements.push(...)` could cause the vector to require growing and reallocating moving all the inner settlements to new memory locations and invalidating this `HashSet`.

How it is right now, its currently fine (since we only modify the `settlements: &mut Vec<Settlement>` after we are done modifying the `HashSet`) but maybe we should be more explicit and maybe capture `settlements` reference preventing mutation for as long as `HashSet` is mutable to the compiler helps you avoid this foot-gun?

Reply:
I pushed another commit to deal with that issue. I added an inner function which takes `&[Settlement]` and returns `HashSet<*const Settlement`. Invalidating the pointers within this function should now be caught by the compiler.
A programmer could still mutate the `&mut settlements` right after finding all settlements but I added a comment there to not do that.